From 45a29d1d850f597e4e0f1687f718bd46ade1de68 Mon Sep 17 00:00:00 2001
From: Samuel Zhang <guoqing.zhang@amd.com>
Date: Thu, 10 Jul 2025 14:23:12 +0800
Subject: [PATCH 026/188] PM: hibernate: add new api
 pm_hibernate_is_recovering()

dev_pm_ops.thaw() is called in following cases:
* normal case: after hibernation image has been created.
* error case 1: creation of a hibernation image has failed.
* error case 2: restoration from a hibernation image has failed.

For normal case, it is called mainly for resume storage devices for
saving the hibernation image. Other devices that are not involved
in the image saving do not need to resume the device. But since there's
no api to know which case thaw() is called, device drivers can't
conditionally resume device in thaw().

The new pm_hibernate_is_recovering() is such a api to query if thaw() is
called in normal case.

Signed-off-by: Samuel Zhang <guoqing.zhang@amd.com>
Acked-by: Rafael J. Wysocki <rafael@kernel.org>
Link: https://lore.kernel.org/r/20250710062313.3226149-5-guoqing.zhang@amd.com
Signed-off-by: Mario Limonciello <mario.limonciello@amd.com>
(cherry picked from commit c2aaddbd2deded9d3301f1bafed242a0f71baba8)
Upstream-Url: git@ssh.gitlab.freedesktop.org:drm/misc/kernel.git
Signed-off-by: Robert Beckett <bob.beckett@collabora.com>
---
 include/linux/suspend.h | 2 ++
 1 file changed, 2 insertions(+)

diff --git a/include/linux/suspend.h b/include/linux/suspend.h
index b02876f1a..7859a5078 100644
--- a/include/linux/suspend.h
+++ b/include/linux/suspend.h
@@ -432,6 +432,8 @@ int is_hibernate_resume_dev(dev_t dev);
 static inline int is_hibernate_resume_dev(dev_t dev) { return 0; }
 #endif
 
+bool pm_hibernate_is_recovering(void);
+
 /* Hibernation and suspend events */
 #define PM_HIBERNATION_PREPARE	0x0001 /* Going to hibernate */
 #define PM_POST_HIBERNATION	0x0002 /* Hibernation finished */
-- 
2.55.0

